numba/lowering.py

```diff
@@ -644,9 +645,17 @@ def lower_call(self, resty, expr):
                                                          argvals, fnty.cconv)
 
         elif isinstance(fnty, types.RecursiveCall):
-            # Self-recursive call
-            impl = imputils.user_function(self.fndesc, ())
-            res = impl(self.context, self.builder, signature, argvals)
+            # Recursive call
+            qualprefix = fnty.overloads[signature.args]
+            mangler = self.context.mangler or default_mangler
+            mangled_name = mangler(qualprefix, signature.args)
+            # special case self recursion
+            if self.builder.function.name.startswith(mangled_name):
+                res = self.context.call_internal(self.builder, self.fndesc,
+                                                 signature, argvals)
+            else:
+                res = self.context.call_unresolved(self.builder, mangled_name,
```

Comment:
It's for the mutual recursion case:

```python
@njit
def foo():             
    ...  # calls bar()

@njit
def bar():
     ... # calls foo()


foo()   
```

When `foo()` is called, the compilation of `bar()` is fully completed (codegen'ed and loaded) before `foo()` is.  Since MCJIT's eager compilation doesn't allow loading modules with declare-only functions (which is needed for `foo()` in `bar()`), the `call_unresolved` injects a global variable that the "linker" can update even after the module is loaded by MCJIT.  The linker would allocate space for the global variable before the `bar()` module is loaded.  When later `foo()` module is defined, it will update `bar()`'s reference to `foo()`.

The legacy lazy JIT and the new ORC JIT would allow a declare-only function be used in a module as long as it is defined by the time of its first use.